src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -140,81 +141,132 @@ initialScannerState = ScannerState
+    process
       :: Maybe PosInfo
       -> NodeType
-      -> [ScannerM Node]
-      -> ScannerM Node
-    remove pos ty subs = do
-      let node = Node pos ty []
-      scan <- use ssIgnore >>= \e -> do
+      -> [ScannerM C.Node]
+      -> ScannerM C.Node
+    process pos ty subs = do
+      let node = C.Node pos ty []
+      use ssIgnore >>= \ign -> do
         -- When no `Ignore` state is set check next node for annotation,
         -- if found then set it as new `IgnoreMode` otherwise skip node.
         let mbAnnotation = getAnnotation node
         case mbAnnotation of
           Just ann -> handleAnnotation pos ty ann
-          Nothing -> case e of
-            Nothing                    -> Node pos ty <$> sequence subs
-            Just (Ignore mode modePos) ->
-                case (mode, ty) of
-                  -- We expect to find a paragraph immediately after the
-                  -- `ignore paragraph` annotanion. If the paragraph is not
-                  -- found we should report an error.
-                  (IMSParagraph, PARAGRAPH)    -> (ssIgnore .= Nothing) $> defNode
-                  (IMSParagraph, x)            -> do
-                    lift . tell . makeError modePos fp . ParagraphErr $ prettyType x
-                    ssIgnore .= Nothing
-                    Node pos ty <$> sequence subs
-
-                  (IMSLink _, LINK {})         -> do
-                    ssIgnore .= Nothing
-                    return defNode
-                  (IMSLink _, IMAGE {})        -> do
-                    ssIgnore .= Nothing
-                    return defNode
-                  (IMSLink ignoreLinkState, _) -> do
-                    when (ignoreLinkState == ExpectingLinkInSubnodes) $
-                      ssIgnore . _Just . ignoreMode .=  IMSLink ParentExpectsLink
-                    node' <- Node pos ty <$> sequence subs
-                    when (ignoreLinkState == ExpectingLinkInSubnodes) $ do
-                      currentIgnore <- use ssIgnore
-                      case currentIgnore of
-                        Just (Ignore {_ignoreMode = IMSLink ParentExpectsLink}) -> do
-                          lift $ tell $ makeError modePos fp LinkErr
-                          ssIgnore .= Nothing
-                        _ -> pass
-                    return node'
-
-      when (ty == PARAGRAPH) $ use ssIgnore >>= \case
+          Nothing -> do
+            case ty of
+              PARAGRAPH -> handleParagraph ign pos ty subs
+              LINK {}   -> handleLink      ign pos ty subs
+              IMAGE {}  -> handleLink      ign pos ty subs
+              _         -> handleOther     ign pos ty subs
+
+    handleLink ::
+      Maybe Ignore ->
+      Maybe PosInfo ->
+      NodeType ->
+      [ScannerM C.Node] ->
+      ScannerM C.Node
+    handleLink ign pos ty subs = do
+      let traverseChildren = C.Node pos ty <$> sequence subs
+      -- It can be checked that it's correct for all the cases
+      ssIgnore .= Nothing
```

Comment:
> Also, AFAIU in the case of ign = IMSParagraph you rely on the behaviour of handleParagraph to strip the entire Paragraph subtree.

It seems that this code is fully made up of such implicit dependencies, because the main reason is how we use annotations. We put an annotation somewhere in the text, and then the further behavior depends on the next nodes.

Reply:
In case of `IMSParagraph`, at least, we could try to look at the next node (not very easy to do) and perform some actions immediately, so, if the next is paragraph, ignore it, and if the next is something else, emit an error. But in case of link annotations we have to handle all this implicit stuff (and that's sad).